File: Python/0441-e-arranging-coins.py
"""
Binary Search
Time: O(logn)
Space: O(1)
"""
class Solution:
    def arrangeCoins(self, n: int) -> int:
        start, end = 1, n
        while start + 1 < end:
            mid = start + (end - start) // 2
            if mid + mid ** 2 <= 2 * n:
                start = mid
            else:
                end = mid
        
        if start + start ** 2 <= 2 * n:
            return start
        return end

File: leetcode/Java/0441-e-arranging-coins.java
/*
 * Binary Search
 */
class Solution {
    public int arrangeCoins(int n) {
        int start = 1, end = n;
        int mid;
        long N = n;
        long cnt;
        while (start + 1 < end) {
            mid = start + (end - start) / 2;
            cnt = (long) (mid + 1) * mid / 2;
            if (cnt <= N)
                start = mid;
            else
                end = mid;
        }
        
        if (start + start * start <= 2 * n)
            return start;
        return end;
    }
}